flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1405,59 +1406,80 @@ private void declineCheckpoint(
+            NotifyCheckpointOperation notifyCheckpointOperation) {
+        TaskInvokable invokable = this.invokable;
+
+        if (executionState == ExecutionState.RUNNING && invokable != null) {
             checkState(invokable instanceof CheckpointableTask, "invokable is not checkpointable");
             try {
-                ((CheckpointableTask) invokable)
-                        .notifyCheckpointAbortAsync(checkpointID, latestCompletedCheckpointId);
+                switch (notifyCheckpointOperation) {
+                    case ABORT:
+                        ((CheckpointableTask) invokable)
+                                .notifyCheckpointAbortAsync(
+                                        checkpointId, latestCompletedCheckpointId);
+                        break;
+                    case COMPLETE:
+                        ((CheckpointableTask) invokable)
+                                .notifyCheckpointCompleteAsync(checkpointId);
+                        break;
+                    case SUBSUME:
+                        ((CheckpointableTask) invokable)
+                                .notifyCheckpointSubsumedAsync(checkpointId);
+                }
             } catch (RejectedExecutionException ex) {
                 // This may happen if the mailbox is closed. It means that the task is shutting
                 // down, so we just ignore it.
                 LOG.debug(
-                        "Notify checkpoint abort {} for {} ({}) was rejected by the mailbox",
-                        checkpointID,
+                        "Notify checkpoint {}} {} for {} ({}) was rejected by the mailbox.",
+                        notifyCheckpointOperation,
+                        checkpointId,
                         taskNameWithSubtask,
                         executionId);
             } catch (Throwable t) {
-                if (getExecutionState() == ExecutionState.RUNNING) {
-                    // fail task if checkpoint aborted notification failed.
-                    failExternally(new RuntimeException("Error while aborting checkpoint", t));
+                switch (notifyCheckpointOperation) {
+                    case ABORT:
+                    case COMPLETE:
+                        if (getExecutionState() == ExecutionState.RUNNING) {
+                            failExternally(
+                                    new RuntimeException(
+                                            String.format(
+                                                    "Error while notify checkpoint %s.",
+                                                    notifyCheckpointOperation),
+                                            t));
+                        }
+                        break;
+                    case SUBSUME:
+                        // just rethrow the throwable out as we do not expect notification of
+                        // subsume could fail the task.
+                        ExceptionUtils.rethrow(t);
```

Comment:
Why do we need a special handling here?

What do you mean "We do not expect notification could fail the task"? Do you mean that this error should never happen? If so, I would keep the error handling unified regardless.

Reply:
From my point of view, we need to ensure task failover directly on the case of checkpoint complete notification to avoid unexpected data lost. However, I don't think we should also apply this rule on checkpoint subsumed notification as there is no strict data completeness request here. WDYT?

Reply:
I'm a little bit thorn apart here.

Wouldn't the same logic apply to `notifyCheckpointAbort()`? 

On the other hand, even in the case of an exception during `notifyCheckpointComplete()`. As long as checkpoint is failed, there won't be any data loss. Data will be committed in the next checkpoint OR if there are enough checkpoint failures, the whole job will failover because of the exceeded number of tolerable checkpoint failures.

I think it could be argued both directions for either of those methods (abort/complete/subsume), so maybe for the sake of simplicity and consistency it would be best to keep them behaving in the same way?

Reply:
I see, for the simplicity I will refactor this to keep the same behavior.

Reply:
Now I remembered some past discussion about the behaviour on failing tasks or not during checkpointing. Our previous conclusion was indeed no to fail the job if not needed. We have to fail the job, if the exception happened in for example the sync phase of snapshotting, as otherwise operators/statebackends/user functions might end up in an inconsistent state. However failures from for example async phase can be handled by the `CheckpointCoordinator` and `CheckpointFailureManager`.

`notifyCheckpointComplete` and `notifyCheckpointAborted` has to fail the job, at least in some/most cases, because if the failure has originated in the user code, that user code might be in an invalid state. For example:
```
void notifyCheckpointComplete() {
  a++;
  maybeThrowException();
  b++;
}
```
after throwing `a` and `b` will be inconsistent. 

As we don't expose `notifyCheckpointSubsumed()` to the users, this argument doesn't hold here. So the only question is our internal data structures are expected to keep working just fine if an exception indeed happens in this call?

Reply:
From my point of view, our internal data structures are expected to keep working fine if come across an exception here. Some customized keyed state backend might introduce IO operations within `notifyCheckpointSubsumed()`, and the task should not fail as those IO operations are best-effort-to-do. That's why my implementation would not fail here.

Reply:
👍 
So let's keep it as it is. (failure of `notifyCheckpointSubsumed` shouldn't cause the task to failover automatically).